File: 02_Object_Oriented_Programming/02_Class_Relationships/01_Inheritance/java/Animal.java
/**
 * Inheritance — Base class (superclass).
 *
 * Key concept: A Dog IS-A Animal.
 * The child class inherits fields & methods from the parent.
 */
public class Animal {

    private String name;

    public Animal() {
        this.name = "Unknown";
    }

    public Animal(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void eat() {
        System.out.println(name + " is eating.");
    }

    public void sleep() {
        System.out.println(name + " is sleeping.");
    }

    @Override
    public String toString() {
        return "Animal(" + name + ")";
    }
}

File: 02_Object_Oriented_Programming/02_Class_Relationships/01_Inheritance/javascript/animal.js
/**
 * inheritance — base class (superclass) in JavaScript.
 *
 * Key concept: A Dog IS-A Animal.
 * The child class inherits fields & methods from the parent.
 */

class Animal {
  #name;

  constructor(name = "Unknown") {
    this.#name = name;
  }

  get name() {
    return this.#name;
  }

  eat() {
    console.log(`${this.#name} is eating.`);
  }

  sleep() {
    console.log(`${this.#name} is sleeping.`);
  }

  toString() {
    return `Animal(${this.#name})`;
  }
}

module.exports = Animal;
